Star in patterns

```
A.scala:4: error: not found: type _*
  val List(1, 2, rest: _*) = List(1, 2, 3, 4, 5) // was
                       ^
exception occured while typechecking A.scala
Exception in thread "main" java.lang.ClassCastException: _$times is not a term name
    at dotty.tools.dotc.core.Names$TypeName.asTermName(Names.scala:184)
    at dotty.tools.dotc.core.Names$TypeName.asTermName(Names.scala:178)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedBind$1.apply(Typer.scala:738)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedBind$1.apply(Typer.scala:735)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedBind(Typer.scala:735)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:905)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:958)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.apply(Typer.scala:252)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedIdent$1.apply(Typer.scala:77)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedIdent(Typer.scala:77)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:902)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:958)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.typedType(Typer.scala:1007)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedTyped$1.regularTyped$1(Typer.scala:328)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedTyped$1.apply(Typer.scala:336)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedTyped$1.apply(Typer.scala:326)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedTyped(Typer.scala:326)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:926)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:960)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedBind$1.apply(Typer.scala:736)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedBind$1.apply(Typer.scala:735)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedBind(Typer.scala:735)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:905)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:958)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedTyped$1.apply(Typer.scala:339)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedTyped$1.apply(Typer.scala:326)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedTyped(Typer.scala:326)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:926)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:960)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Applications$$anonfun$typedUnApply$1$$anonfun$15.apply(Applications.scala:696)
    at dotty.tools.dotc.typer.Applications$$anonfun$typedUnApply$1$$anonfun$15.apply(Applications.scala:696)
    at scala.runtime.Tuple2Zipped$$anonfun$map$extension$1.apply(Tuple2Zipped.scala:42)
    at scala.runtime.Tuple2Zipped$$anonfun$map$extension$1.apply(Tuple2Zipped.scala:40)
    at scala.collection.immutable.List.foreach(List.scala:383)
    at scala.runtime.Tuple2Zipped$.map$extension(Tuple2Zipped.scala:40)
    at dotty.tools.dotc.typer.Applications$$anonfun$typedUnApply$1.apply(Applications.scala:696)
    at dotty.tools.dotc.typer.Applications$$anonfun$typedUnApply$1.apply(Applications.scala:529)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Applications$class.typedUnApply(Applications.scala:529)
    at dotty.tools.dotc.typer.Typer.typedUnApply(Typer.scala:56)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:921)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:960)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.typedPattern(Typer.scala:1009)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$dotty$tools$dotc$typer$Typer$$anonfun$$typedCase$1$1$$anonfun$12.apply(Typer.scala:596)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$dotty$tools$dotc$typer$Typer$$anonfun$$typedCase$1$1$$anonfun$12.apply(Typer.scala:596)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$dotty$tools$dotc$typer$Typer$$anonfun$$typedCase$1$1.apply(Typer.scala:597)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$dotty$tools$dotc$typer$Typer$$anonfun$$typedCase$1$1.apply(Typer.scala:582)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1.dotty$tools$dotc$typer$Typer$$anonfun$$typedCase$1(Typer.scala:582)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$13.apply(Typer.scala:600)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1$$anonfun$13.apply(Typer.scala:600)
    at dotty.tools.dotc.core.Decorators$ListDecorator$.loop$1(Decorators.scala:50)
    at dotty.tools.dotc.core.Decorators$ListDecorator$.mapconserve$extension(Decorators.scala:66)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1.apply(Typer.scala:600)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedMatch$1.apply(Typer.scala:554)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedMatch(Typer.scala:553)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:933)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:960)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1005)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:769)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedValDef$1.apply(Typer.scala:763)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedValDef(Typer.scala:763)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:908)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:958)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:989)
    at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1001)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:821)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:793)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:793)
    at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:913)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:958)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.traverse$1(Typer.scala:989)
    at dotty.tools.dotc.typer.Typer.typedStats(Typer.scala:1001)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:854)
    at dotty.tools.dotc.typer.Typer$$anonfun$typedPackageDef$1.apply(Typer.scala:845)
    at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
    at dotty.tools.dotc.typer.Typer.typedPackageDef(Typer.scala:845)
    at dotty.tools.dotc.typer.Typer.typedUnnamed$1(Typer.scala:949)
    at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:960)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:967)
    at dotty.tools.dotc.typer.Typer$$anonfun$typed$2.apply(Typer.scala:965)
    at dotty.tools.dotc.reporting.Reporting$class.traceIndented(Reporter.scala:145)
    at dotty.tools.dotc.core.Contexts$Context.traceIndented(Contexts.scala:50)
    at dotty.tools.dotc.typer.Typer.typed(Typer.scala:965)
    at dotty.tools.dotc.typer.Typer.typedExpr(Typer.scala:1005)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$typeCheck$1.apply$mcV$sp(FrontEnd.scala:37)
    at dotty.tools.dotc.typer.FrontEnd.monitor(FrontEnd.scala:16)
    at dotty.tools.dotc.typer.FrontEnd.typeCheck(FrontEnd.scala:35)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:48)
    at dotty.tools.dotc.typer.FrontEnd$$anonfun$runOn$3.apply(FrontEnd.scala:48)
    at scala.collection.immutable.List.foreach(List.scala:383)
    at dotty.tools.dotc.typer.FrontEnd.runOn(FrontEnd.scala:48)
    at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:42)
    at dotty.tools.dotc.Run$$anonfun$compileSources$1$$anonfun$apply$mcV$sp$3.apply(Run.scala:40)
    at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
    at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
    at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply$mcV$sp(Run.scala:40)
    at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:32)
    at dotty.tools.dotc.Run$$anonfun$compileSources$1.apply(Run.scala:32)
    at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:68)
    at dotty.tools.dotc.Run.compileSources(Run.scala:31)
    at dotty.tools.dotc.Run.compile(Run.scala:28)
    at dotty.tools.dotc.Driver.doCompile(Driver.scala:19)
    at dotty.tools.dotc.Main$.doCompile(Main.scala:26)
    at dotty.tools.dotc.Driver.process(Driver.scala:30)
    at dotty.tools.dotc.Driver.main(Driver.scala:44)
    at dotty.tools.dotc.Main.main(Main.scala)
```

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -860,7 +860,7 @@ object desugar {
       case tree @ Bind(_, tree1) =>
         add(tree, TypeTree())
         collect(tree1)
-      case Typed(id: Ident, t) if isVarPattern(id) && id.name != nme.WILDCARD =>
+      case Typed(id: Ident, t) if isVarPattern(id) && id.name != nme.WILDCARD && !isWildcardStarArg(tree) =>
         add(id, t)
       case id: Ident if isVarPattern(id) && id.name != nme.WILDCARD =>
         add(id, TypeTree())

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -325,7 +325,11 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
 
   def typedTyped(tree: untpd.Typed, pt: Type)(implicit ctx: Context): Tree = track("typedTyped") {
     def regularTyped(isWildcard: Boolean) = {
-      val tpt1 = typedType(tree.tpt)
+      val tpt1 =
+        if (untpd.isWildcardStarArg(tree))
+          TypeTree(defn.SeqClass.typeRef.appliedTo(pt :: Nil))
+        else
+          typedType(tree.tpt)
       val expr1 =
         if (isWildcard) tree.expr withType tpt1.tpe
         else typed(tree.expr, tpt1.tpe)
@@ -336,7 +340,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         if (id.name == nme.WILDCARD) regularTyped(isWildcard = true)
         else {
           import untpd._
-          typed(Bind(id.name, Typed(Ident(nme.WILDCARD), tree.tpt)).withPos(id.pos))
+          typed(Bind(id.name, Typed(Ident(nme.WILDCARD), tree.tpt)).withPos(id.pos), pt)
         }
       case _ =>
         if (untpd.isWildcardStarArg(tree))

--- a/tests/pos/vararg-pattern.scala
+++ b/tests/pos/vararg-pattern.scala
@@ -1,0 +1,12 @@
+object Test {
+
+  List(1, 2, 3, 4) match {
+    case List(1, 2, xs: _*) =>
+      val ys: Seq[Int] = xs
+      println(ys)
+  }
+  val List(1, 2, x: _*) = List(1, 2, 3, 4)
+
+}
+
+

